modules/git/git.go

```diff
@@ -165,46 +164,46 @@ func InitSimple(ctx context.Context) error {
-// otherwise there will be data-race problem at the moment.
-func InitOnceWithSync(ctx context.Context) (err error) {
+// InitFull initializes git module with version check and change global variables, sync gitconfig.
+// It should only be called once at the beginning of the program initialization (TestMain/GlobalInitInstalled) as this code makes unsynchronized changes to variables.
+func InitFull(ctx context.Context) (err error) {
 	if err = checkInit(); err != nil {
 		return err
 	}
 
-	initOnce.Do(func() {
-		if err = InitSimple(ctx); err != nil {
-			return
-		}
+	if err = InitSimple(ctx); err != nil {
+		return
+	}
 
-		// when git works with gnupg (commit signing), there should be a stable home for gnupg commands
-		if _, ok := os.LookupEnv("GNUPGHOME"); !ok {
-			_ = os.Setenv("GNUPGHOME", filepath.Join(HomeDir(), ".gnupg"))
-		}
+	// when git works with gnupg (commit signing), there should be a stable home for gnupg commands
+	if _, ok := os.LookupEnv("GNUPGHOME"); !ok {
+		_ = os.Setenv("GNUPGHOME", filepath.Join(HomeDir(), ".gnupg"))
+	}
 
-		// Since git wire protocol has been released from git v2.18
-		if setting.Git.EnableAutoGitWireProtocol && CheckGitVersionAtLeast("2.18") == nil {
-			globalCommandArgs = append(globalCommandArgs, "-c", "protocol.version=2")
-		}
+	// Since git wire protocol has been released from git v2.18
+	if setting.Git.EnableAutoGitWireProtocol && CheckGitVersionAtLeast("2.18") == nil {
+		globalCommandArgs = append(globalCommandArgs, "-c", "protocol.version=2")
+	}
 
-		// By default partial clones are disabled, enable them from git v2.22
-		if !setting.Git.DisablePartialClone && CheckGitVersionAtLeast("2.22") == nil {
-			globalCommandArgs = append(globalCommandArgs, "-c", "uploadpack.allowfilter=true", "-c", "uploadpack.allowAnySHA1InWant=true")
-		}
+	// By default partial clones are disabled, enable them from git v2.22
+	if !setting.Git.DisablePartialClone && CheckGitVersionAtLeast("2.22") == nil {
+		globalCommandArgs = append(globalCommandArgs, "-c", "uploadpack.allowfilter=true", "-c", "uploadpack.allowAnySHA1InWant=true")
+	}
 
-		// Explicitly disable credential helper, otherwise Git credentials might leak
-		if CheckGitVersionAtLeast("2.9") == nil {
-			globalCommandArgs = append(globalCommandArgs, "-c", "credential.helper=")
-		}
+	// Explicitly disable credential helper, otherwise Git credentials might leak
+	if CheckGitVersionAtLeast("2.9") == nil {
+		globalCommandArgs = append(globalCommandArgs, "-c", "credential.helper=")
+	}
 
-		SupportProcReceive = CheckGitVersionAtLeast("2.29") == nil
-	})
-	if err != nil {
-		return err
+	SupportProcReceive = CheckGitVersionAtLeast("2.29") == nil
+
+	if setting.LFS.StartServer {
+		if CheckGitVersionAtLeast("2.1.2") != nil {
+			return errors.New("LFS server support requires Git >= 2.1.2")
+		}
+		globalCommandArgs = append(globalCommandArgs, "-c", "filter.lfs.required=", "-c", "filter.lfs.smudge=", "-c", "filter.lfs.clean=")
```

Comment:
Also, does `append` work on `nil`?
Because you explicitly set it to `nil` in `InitSimple` (for whatever reason?), and never actually assign it afterward?

Reply:
Why doesn't `append` work on nil .... can you check Golang's syntax and manual first .......

The reason is in comment:  `InitSimple initializes git module with a very simple step, no config changes, no global command arguments.`

Reply:
Interesting, didn't know that `append` initializes `null` slices.